### PR TITLE
Issue 2798: Add precondition checks on splits and merges from the controller.

### LINKED ISSUE #2798: Add more assertions about splits and merges.
**Problem description**
The client should double check that the controller is sending it valid data when it asks for successors.

**Problem location**
StreamSegments

**Suggestions for an improvement**
Add some precondition checks to make sure the data is plausibly valid.

### PR DESCRIPTION
**Pull request by tkaitchuck**
**Change log description**
Add some precondition checks on StreamSegments returned from the controller.

**Purpose of the change**
Fixes #2798 

**What the code does**
Adds some sanity checks for the data coming from the controller after a segment sealed.

**How to verify it**
Existing tests actually cover this fairly well.

### CODE REVIEW

**Comment by shrids** [comment]
Travis build failed with this test failure, will re-trigger the build.
```
io.pravega.segmentstore.server.logs.DurableLogTests > testTruncateWithoutRecovery FAILED
    java.lang.AssertionError: Unexpected number of operations added after full truncation. expected:<3> but was:<13>
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.failNotEquals(Assert.java:834)
        at org.junit.Assert.assertEquals(Assert.java:645)
        at io.pravega.segmentstore.server.logs.DurableLogTests.testTruncateWithoutRecovery(DurableLogTests.java:1191)```

File: client/src/main/java/io/pravega/client/stream/impl/ControllerImpl.java [hunk 1]
```diff
@@ -693,6 +693,7 @@ private Boolean handleScaleResponse(Stream stream, ScaleResponse response) {
             log.debug("Received the following data from the controller {}", ranges.getSegmentRangesList());
             NavigableMap<Double, Segment> rangeMap = new TreeMap<>();
             for (SegmentRange r : ranges.getSegmentRangesList()) {
+                Preconditions.checkState(r.getMinKey() <= r.getMaxKey());
```
Comment (shrids) [inline client/src/main/java/io/pravega/client/stream/impl/ControllerImpl.java]: We could add an appropriate error message for the added ` Preconditions.checkState` and `Preconditions.checkArgument`

File: client/src/main/java/io/pravega/client/stream/impl/ControllerImpl.java [hunk 2]
```diff
@@ -693,6 +693,9 @@ private Boolean handleScaleResponse(Stream stream, ScaleResponse response) {
             log.debug("Received the following data from the controller {}", ranges.getSegmentRangesList());
             NavigableMap<Double, Segment> rangeMap = new TreeMap<>();
             for (SegmentRange r : ranges.getSegmentRangesList()) {
+                Preconditions.checkState(r.getMinKey() <= r.getMaxKey(),
```
Comment (shiveshr) [inline client/src/main/java/io/pravega/client/stream/impl/ControllerImpl.java]: shouldnt this be strictly less than?

Comment (tkaitchuck) [reply client/src/main/java/io/pravega/client/stream/impl/ControllerImpl.java]: Well in theory it is possible for a segment to be split so far that it extends to the maximum precision of a double and is left with nothing in one half... But that should not occur in practice.

**Pull request merged**